arwen/contexts/async.go

```diff
@@ -82,34 +80,77 @@ func NewAsyncContext(
 
 // InitState initializes the internal state of the AsyncContext.
 func (context *asyncContext) InitState() {
+	context.address = nil
+	context.callID = nil
+	context.callerCallID = nil
 	context.callerAddr = make([]byte, 0)
-	context.gasPrice = 0
 	context.gasAccumulated = 0
 	context.returnData = make([]byte, 0)
 	context.asyncCallGroups = make([]*arwen.AsyncCallGroup, 0)
 	context.callback = ""
+	context.callbackData = make([]byte, 0)
+	context.callbackAsyncInitiatorCallID = nil
+	context.callsCounter = 0
+	context.totalCallsCounter = 0
+	context.childResults = nil
 }
 
 // InitStateFromInput initializes the internal state of the AsyncContext with
 // information provided by a ContractCallInput.
 func (context *asyncContext) InitStateFromInput(input *vmcommon.ContractCallInput) {
 	context.InitState()
 	context.callerAddr = input.CallerAddr
-	context.gasPrice = input.GasPrice
-	context.gasAccumulated = 0
+	context.callType = input.CallType
+
+	runtime := context.host.Runtime()
+	context.address = runtime.GetSCAddress()
+
+	emptyStack := len(context.stateStack) == 0
+	if emptyStack && !context.isCallAsync() {
+		context.callID = input.CurrentTxHash
+		context.callerCallID = nil
+	} else {
+		context.callID = runtime.PopFirstArgumentFromVMInput()
+		context.callerCallID = runtime.PopFirstArgumentFromVMInput()
+	}
+
+	if input.CallType == vm.AsynchronousCallBack {
+		context.callbackAsyncInitiatorCallID = runtime.PopFirstArgumentFromVMInput()
+		context.gasAccumulated = big.NewInt(0).SetBytes(runtime.PopFirstArgumentFromVMInput()).Uint64()
```

Comment:
no part of gas accumulated will end up in some gas provided

Reply:
As @matei-p said, `context.gasAccumulated` will return back to the caller in `vmOutput.GasRemaining`. It will never be used for execution anymore, it will never be added to some `vmInput.GasProvided`.